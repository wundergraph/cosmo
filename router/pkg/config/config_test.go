package config

import (
	"os"
	"regexp"
	"testing"
	"time"

	"github.com/santhosh-tekuri/jsonschema/v5"
	"github.com/sebdah/goldie/v2"
	"github.com/stretchr/testify/require"
)

func TestConfigRequiredValues(t *testing.T) {
	_, err := LoadConfig("./fixtures/with_required_values.yaml", "")
	require.ErrorContains(t, err, "either router config path or graph token must be provided")
}

func TestTokenNotRequiredWhenPassingStaticConfig(t *testing.T) {
	_, err := LoadConfig("./fixtures/with_static_execution_config.yaml", "")

	require.NoError(t, err)
}

func TestCustomBytesExtension(t *testing.T) {
	_, err := LoadConfig("./fixtures/minimum_bytes_error.yaml", "")

	var js *jsonschema.ValidationError
	require.ErrorAs(t, err, &js)

	require.Equal(t, js.Causes[0].KeywordLocation, "/properties/traffic_shaping/properties/router/properties/max_request_body_size/bytes")
	require.Equal(t, js.Causes[0].Message, "must be greater or equal than 1.0 MB")
}

func TestVariableExpansion(t *testing.T) {
	require.NoError(t, os.Setenv("TEST_POLL_INTERVAL", "20s"))

	t.Cleanup(func() {
		require.NoError(t, os.Unsetenv("TEST_POLL_INTERVAL"))
	})

	cfg, err := LoadConfig("./fixtures/variable_expansion.yaml", "")

	require.NoError(t, err)

	require.Equal(t, cfg.Config.PollInterval, time.Second*20)
}

func TestConfigHasPrecedence(t *testing.T) {
	require.NoError(t, os.Setenv("POLL_INTERVAL", "22s"))

	t.Cleanup(func() {
		require.NoError(t, os.Unsetenv("POLL_INTERVAL"))
	})

	cfg, err := LoadConfig("./fixtures/config_precedence.yaml", "")

	require.NoError(t, err)

	require.Equal(t, cfg.Config.PollInterval, time.Second*11)
}

func TestErrorWhenConfigNotExists(t *testing.T) {
	_, err := LoadConfig("./fixtures/not_exists.yaml", "")

	require.Error(t, err)
	require.ErrorContains(t, err, "could not read custom config file ./fixtures/not_exists.yaml: open ./fixtures/not_exists.yaml: no such file or directory")
}

func TestRegexDecoding(t *testing.T) {
	cfg, err := LoadConfig("./fixtures/regex-empty-decoding.yaml", "")

	require.NoError(t, err)
	require.Len(t, cfg.Config.Telemetry.Metrics.Prometheus.ExcludeMetrics, 0)
	require.Len(t, cfg.Config.Telemetry.Metrics.Prometheus.ExcludeMetricLabels, 0)

	cfg, err = LoadConfig("./fixtures/regex-decoding.yaml", "")

	require.NoError(t, err)
	require.Len(t, cfg.Config.Telemetry.Metrics.Prometheus.ExcludeMetrics, 2)
	require.Len(t, cfg.Config.Telemetry.Metrics.Prometheus.ExcludeMetricLabels, 1)
	require.Equal(t, cfg.Config.Telemetry.Metrics.Prometheus.ExcludeMetrics, RegExArray{regexp.MustCompile("^go_.*"), regexp.MustCompile("^process_.*")})
	require.Equal(t, cfg.Config.Telemetry.Metrics.Prometheus.ExcludeMetricLabels, RegExArray{regexp.MustCompile("^instance")})
}

func TestErrorWhenEnvVariableConfigNotExists(t *testing.T) {
	require.NoError(t, os.Setenv("CONFIG_PATH", "not_exists.yaml"))

	t.Cleanup(func() {
		require.NoError(t, os.Unsetenv("CONFIG_PATH"))
	})

	_, err := LoadConfig("", "")

	require.Error(t, err)
	require.ErrorContains(t, err, "could not read custom config file not_exists.yaml: open not_exists.yaml: no such file or directory")
}

func TestConfigIsOptional(t *testing.T) {

	require.NoError(t, os.Setenv("GRAPH_API_TOKEN", "XXX"))

	t.Cleanup(func() {
		require.NoError(t, os.Unsetenv("GRAPH_API_TOKEN"))
	})

	result, err := LoadConfig("", "")

	require.NoError(t, err)
	require.False(t, result.DefaultLoaded)
}

func TestCustomGoDurationExtension(t *testing.T) {
	_, err := LoadConfig("./fixtures/min_duration_error.yaml", "")

	var js *jsonschema.ValidationError
	require.ErrorAs(t, err, &js)

	require.Equal(t, js.Causes[0].KeywordLocation, "/properties/telemetry/properties/tracing/properties/exporters/items/properties/export_timeout/duration")
	require.Equal(t, js.Causes[0].Message, "must be greater or equal than 5s")

	_, err = LoadConfig("./fixtures/max_duration_error.yaml", "")

	require.ErrorAs(t, err, &js)

	require.Equal(t, js.Causes[0].KeywordLocation, "/properties/telemetry/properties/tracing/properties/exporters/items/properties/export_timeout/duration")
	require.Equal(t, js.Causes[0].Message, "must be less or equal than 2m0s")
}

func TestLoadFullConfig(t *testing.T) {
	cfg, err := LoadConfig("./fixtures/full.yaml", "")
	require.NoError(t, err)

	g := goldie.New(
		t,
		goldie.WithFixtureDir("testdata"),
		goldie.WithNameSuffix(".json"),
		goldie.WithDiffEngine(goldie.ClassicDiff),
	)

	g.AssertJson(t, "config_full", cfg.Config)
}

func TestDefaults(t *testing.T) {
	// Set in the CI to false. We need to unset it to test the default values
	_ = os.Unsetenv("ROUTER_REGISTRATION")

	cfg, err := LoadConfig("./fixtures/minimal.yaml", "")
	require.NoError(t, err)

	g := goldie.New(
		t,
		goldie.WithFixtureDir("testdata"),
		goldie.WithNameSuffix(".json"),
		goldie.WithDiffEngine(goldie.ClassicDiff),
	)

	g.AssertJson(t, "config_defaults", cfg.Config)
}

func TestEventsInvalidProvider(t *testing.T) {
	_, err := LoadConfig("./fixtures/events/invalid_provider.yaml", "")
	// Note: If none of the oneOf array matches, the first in the array is compared
	require.ErrorContains(t, err, "missing properties: 'url'")
}

func TestInvalidAuthenticatedNatsProviderNoToken(t *testing.T) {
	_, err := LoadConfig("./fixtures/events/invalid_authenticated_nats_provider_no_token.yaml", "")
	// Note: If none of the oneOf array matches, the first in the array is compared
	require.ErrorContains(t, err, "missing properties: 'token'")
}

func TestInvalidAuthenticatedNatsProviderNoUsername(t *testing.T) {
	_, err := LoadConfig("./fixtures/events/invalid_authenticated_nats_provider_no_username.yaml", "")
	// Note: If none of the oneOf array matches, the first in the array is compared
	require.ErrorContains(t, err, "missing properties: 'token'")
}

func TestInvalidAuthenticatedNatsProviderNoPassword(t *testing.T) {
	_, err := LoadConfig("./fixtures/events/invalid_authenticated_nats_provider_no_password.yaml", "")
	// Note: If none of the oneOf array matches, the first in the array is compared
	require.ErrorContains(t, err, "missing properties: 'token'")
}

func TestValidUnauthenticatedNatsProvider(t *testing.T) {
	_, err := LoadConfig("./fixtures/events/valid_unauthenticated_nats_provider.yaml", "")
	require.NoError(t, err)
}

func TestValidAuthenticatedNatsProviderWithToken(t *testing.T) {
	_, err := LoadConfig("./fixtures/events/valid_authenticated_nats_provider_with_token.yaml", "")
	require.NoError(t, err)
}

func TestValidAuthenticatedNatsProviderWithUsernamePassword(t *testing.T) {
	_, err := LoadConfig("./fixtures/events/valid_authenticated_nats_provider_with_username_password.yaml", "")
	require.NoError(t, err)
}
