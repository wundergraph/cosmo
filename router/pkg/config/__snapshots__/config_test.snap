
[TestLoadFullConfig - 1]
{
 "Authentication": {
  "Providers": [
   {
    "JWKS": {
     "HeaderNames": [
      "Authorization"
     ],
     "HeaderValuePrefixes": [
      "Bearer"
     ],
     "RefreshInterval": 60000000000,
     "URL": "https://example.com/.well-known/jwks.json"
    },
    "Name": "My Auth Provider"
   }
  ]
 },
 "Authorization": {
  "RejectOperationIfUnauthorized": false,
  "RequireAuthentication": false
 },
 "CDN": {
  "CacheSize": 100000000,
  "URL": "https://cosmo-cdn.wundergraph.com"
 },
 "CORS": {
  "AllowCredentials": true,
  "AllowHeaders": [
   "Origin",
   "Content-Length",
   "Content-Type"
  ],
  "AllowMethods": [
   "HEAD",
   "GET",
   "POST"
  ],
  "AllowOrigins": [
   "*"
  ],
  "MaxAge": 300000000000
 },
 "Cluster": {
  "Name": "my-cluster"
 },
 "ControlplaneURL": "https://cosmo-cp.wundergraph.com",
 "DevelopmentMode": false,
 "EngineExecutionConfiguration": {
  "Debug": {
   "ConfigurationVisitor": false,
   "DatasourceVisitor": false,
   "PlanningVisitor": false,
   "PrintNodeSuggestions": false,
   "PrintOperationEnableASTRefs": false,
   "PrintOperationTransformations": false,
   "PrintPlanningPaths": false,
   "PrintQueryPlans": false,
   "ReportMemoryUsage": false,
   "ReportWebSocketConnections": false
  },
  "EnableExecutionPlanCacheResponseHeader": false,
  "EnableRequestTracing": true,
  "EnableSingleFlight": true,
  "EnableWebSocketEpollKqueue": true,
  "EpollKqueueConnBufferSize": 128,
  "EpollKqueuePollTimeout": 1000000000,
  "ExecutionPlanCacheSize": 10000,
  "MaxConcurrentResolvers": 1024,
  "WebSocketReadTimeout": 1000000000
 },
 "Events": {
  "Sources": [
   {
    "Provider": "NATS",
    "URL": "nats://localhost:4222"
   }
  ]
 },
 "GracePeriod": 20000000000,
 "Graph": {
  "Token": "mytoken"
 },
 "GraphQLPath": "/graphql",
 "GraphqlMetrics": {
  "CollectorEndpoint": "https://cosmo-metrics.wundergraph.com",
  "Enabled": true
 },
 "Headers": {
  "All": {
   "Request": [
    {
     "Default": "",
     "Matching": "",
     "Named": "X-Test-Header",
     "Operation": "propagate"
    },
    {
     "Default": "",
     "Matching": "(?i)^X-Custom-.*",
     "Named": "",
     "Operation": "propagate"
    },
    {
     "Default": "123",
     "Matching": "",
     "Named": "X-User-Id",
     "Operation": "propagate"
    }
   ]
  },
  "Subgraphs": {
   "specific-subgraph": {
    "Request": [
     {
      "Default": "some-secret",
      "Matching": "",
      "Named": "Subgraph-Secret",
      "Operation": "propagate"
     }
    ]
   }
  }
 },
 "HealthCheckPath": "/health",
 "InstanceID": "",
 "IntrospectionEnabled": true,
 "JSONLog": true,
 "ListenAddr": "localhost:3002",
 "LivenessCheckPath": "/health/live",
 "LocalhostFallbackInsideDocker": true,
 "LogLevel": "info",
 "Modules": {
  "myModule": {
   "value": 1
  }
 },
 "OverrideRoutingURL": {
  "Subgraphs": {
   "some-subgraph": "http://router:3002/graphql"
  }
 },
 "PlaygroundEnabled": true,
 "PlaygroundPath": "/",
 "PollInterval": 10000000000,
 "RateLimit": {
  "Debug": false,
  "Enabled": true,
  "SimpleStrategy": {
   "Burst": 60,
   "Period": 60000000000,
   "Rate": 60,
   "RejectExceedingRequests": true
  },
  "Storage": {
   "Addr": "localhost:6379",
   "KeyPrefix": "cosmo_rate_limit",
   "Password": "test"
  },
  "Strategy": "simple"
 },
 "ReadinessCheckPath": "/health/ready",
 "RouterConfigPath": "",
 "RouterRegistration": true,
 "ShutdownDelay": 15000000000,
 "Telemetry": {
  "Metrics": {
   "OTLP": {
    "Enabled": true,
    "Exporters": [
     {
      "Disabled": false,
      "Endpoint": "https://my-otel-collector.example.com",
      "Exporter": "http",
      "HTTPPath": "/v1/metrics",
      "Headers": {}
     }
    ],
    "RouterRuntime": true
   },
   "Prometheus": {
    "Enabled": true,
    "ExcludeMetricLabels": [
     ""
    ],
    "ExcludeMetrics": [
     ""
    ],
    "ListenAddr": "127.0.0.1:8088",
    "Path": "/metrics"
   }
  },
  "ServiceName": "cosmo-router",
  "Tracing": {
   "Enabled": true,
   "ExportGraphQLVariables": true,
   "Exporters": [
    {
     "BatchTimeout": 10000000000,
     "Disabled": false,
     "Endpoint": "https://my-otel-collector.example.com",
     "ExportTimeout": 30000000000,
     "Exporter": "http",
     "HTTPPath": "/v1/traces",
     "Headers": null
    }
   ],
   "Propagation": {
    "B3": false,
    "Baggage": false,
    "Jaeger": false,
    "TraceContext": true
   },
   "SamplingRate": 1,
   "WithNewRoot": false
  }
 },
 "TrafficShaping": {
  "All": {
   "BackoffJitterRetry": {
    "Algorithm": "backoff_jitter",
    "Enabled": true,
    "Interval": 3000000000,
    "MaxAttempts": 5,
    "MaxDuration": 10000000000
   },
   "DialTimeout": 30000000000,
   "ExpectContinueTimeout": 0,
   "KeepAliveIdleTimeout": 0,
   "KeepAliveProbeInterval": 30000000000,
   "RequestTimeout": 60000000000,
   "ResponseHeaderTimeout": 0,
   "TLSHandshakeTimeout": 0
  },
  "Router": {
   "MaxRequestBodyBytes": 5000000
  }
 },
 "Version": "1",
 "WebSocket": {
  "AbsintheProtocol": {
   "Enabled": true,
   "HandlerPath": "/absinthe/socket"
  },
  "Enabled": true,
  "ForwardInitialPayload": true,
  "ForwardUpgradeHeaders": true,
  "ForwardUpgradeQueryParams": true
 }
}
---

[TestDefaults - 1]
{
 "Authentication": {
  "Providers": null
 },
 "Authorization": {
  "RejectOperationIfUnauthorized": false,
  "RequireAuthentication": false
 },
 "CDN": {
  "CacheSize": 100000000,
  "URL": "https://cosmo-cdn.wundergraph.com"
 },
 "CORS": {
  "AllowCredentials": true,
  "AllowHeaders": [
   "Origin",
   "Content-Length",
   "Content-Type"
  ],
  "AllowMethods": [
   "HEAD",
   "GET",
   "POST"
  ],
  "AllowOrigins": [
   "*"
  ],
  "MaxAge": 300000000000
 },
 "Cluster": {
  "Name": ""
 },
 "ControlplaneURL": "https://cosmo-cp.wundergraph.com",
 "DevelopmentMode": false,
 "EngineExecutionConfiguration": {
  "Debug": {
   "ConfigurationVisitor": false,
   "DatasourceVisitor": false,
   "PlanningVisitor": false,
   "PrintNodeSuggestions": false,
   "PrintOperationEnableASTRefs": false,
   "PrintOperationTransformations": false,
   "PrintPlanningPaths": false,
   "PrintQueryPlans": false,
   "ReportMemoryUsage": false,
   "ReportWebSocketConnections": false
  },
  "EnableExecutionPlanCacheResponseHeader": false,
  "EnableRequestTracing": true,
  "EnableSingleFlight": true,
  "EnableWebSocketEpollKqueue": true,
  "EpollKqueueConnBufferSize": 128,
  "EpollKqueuePollTimeout": 1000000000,
  "ExecutionPlanCacheSize": 10000,
  "MaxConcurrentResolvers": 1024,
  "WebSocketReadTimeout": 5000000000
 },
 "Events": {
  "Sources": null
 },
 "GracePeriod": 20000000000,
 "Graph": {
  "Token": "token"
 },
 "GraphQLPath": "/graphql",
 "GraphqlMetrics": {
  "CollectorEndpoint": "https://cosmo-metrics.wundergraph.com",
  "Enabled": true
 },
 "Headers": {
  "All": {
   "Request": null
  },
  "Subgraphs": null
 },
 "HealthCheckPath": "/health",
 "InstanceID": "",
 "IntrospectionEnabled": true,
 "JSONLog": true,
 "ListenAddr": "localhost:3003",
 "LivenessCheckPath": "/health/live",
 "LocalhostFallbackInsideDocker": true,
 "LogLevel": "info",
 "Modules": null,
 "OverrideRoutingURL": {
  "Subgraphs": {}
 },
 "PlaygroundEnabled": true,
 "PlaygroundPath": "/",
 "PollInterval": 10000000000,
 "RateLimit": {
  "Debug": false,
  "Enabled": false,
  "SimpleStrategy": {
   "Burst": 10,
   "Period": 1000000000,
   "Rate": 10,
   "RejectExceedingRequests": false
  },
  "Storage": {
   "Addr": "localhost:6379",
   "KeyPrefix": "cosmo_rate_limit",
   "Password": ""
  },
  "Strategy": "simple"
 },
 "ReadinessCheckPath": "/health/ready",
 "RouterConfigPath": "",
 "RouterRegistration": true,
 "ShutdownDelay": 60000000000,
 "Telemetry": {
  "Metrics": {
   "OTLP": {
    "Enabled": true,
    "Exporters": [],
    "RouterRuntime": true
   },
   "Prometheus": {
    "Enabled": true,
    "ExcludeMetricLabels": null,
    "ExcludeMetrics": null,
    "ListenAddr": "127.0.0.1:8088",
    "Path": "/metrics"
   }
  },
  "ServiceName": "cosmo-router",
  "Tracing": {
   "Enabled": true,
   "ExportGraphQLVariables": false,
   "Exporters": [],
   "Propagation": {
    "B3": false,
    "Baggage": false,
    "Jaeger": false,
    "TraceContext": true
   },
   "SamplingRate": 1,
   "WithNewRoot": false
  }
 },
 "TrafficShaping": {
  "All": {
   "BackoffJitterRetry": {
    "Algorithm": "backoff_jitter",
    "Enabled": true,
    "Interval": 3000000000,
    "MaxAttempts": 5,
    "MaxDuration": 10000000000
   },
   "DialTimeout": 30000000000,
   "ExpectContinueTimeout": 0,
   "KeepAliveIdleTimeout": 0,
   "KeepAliveProbeInterval": 30000000000,
   "RequestTimeout": 60000000000,
   "ResponseHeaderTimeout": 0,
   "TLSHandshakeTimeout": 10000000000
  },
  "Router": {
   "MaxRequestBodyBytes": 5000000
  }
 },
 "Version": "1",
 "WebSocket": {
  "AbsintheProtocol": {
   "Enabled": true,
   "HandlerPath": "/absinthe/socket"
  },
  "Enabled": true,
  "ForwardInitialPayload": true,
  "ForwardUpgradeHeaders": true,
  "ForwardUpgradeQueryParams": true
 }
}
---
